crates/bevy_math/src/primitives/dim3.rs

```diff
@@ -629,6 +629,132 @@ impl Torus {
     }
 }
 
+/// A 3D triangle primitive.
+#[derive(Clone, Copy, Debug, PartialEq)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+pub struct Triangle3d {
+    /// The vertices of the triangle
+    pub vertices: [Vec3; 3],
+}
+
+impl Primitive3d for Triangle3d {}
+
+impl Default for Triangle3d {
+    /// Returns the default [`Triangle3d`] with vertices at the `[-0.5, 0.0, 0.0]`, `[0.5, 0.0, 0.0]`, and `[0.0, 0.5, 0.0]`
+    fn default() -> Self {
+        Self {
+            vertices: [Vec3::ZERO, Vec3::X, Vec3::Y],
+        }
+    }
+}
+
+impl Triangle3d {
+    /// Create a new [`Triangle3d`] from points `a`, `b`, and `c`
+    #[inline(always)]
+    pub fn new(a: Vec3, b: Vec3, c: Vec3) -> Self {
+        Self {
+            vertices: [a, b, c],
+        }
+    }
+
+    /// Get the area of the triangle
+    #[inline(always)]
+    pub fn area(&self) -> f32 {
+        let [a, b, c] = self.vertices;
+        let ab = a.distance(b);
+        let bc = b.distance(c);
+        let ca = c.distance(a);
+
+        let s = (ab + bc + ca) / 2.0;
+        (s * (s - ab) * (s - bc) * (s - ca)).sqrt()
```

Comment:
Heron's formula I see.

However this could be computed much more efficiently via the cross product (skipping 3 sqrts and overall less operations):
```suggestion
        let ab = a - b;
        let ac = a - c;
        // The length of the cross product is `ab.length() * ac.length() * sin(angle_between_ab_and_ac)`
        // Which is the area of the paralelogram formed by `ab` and `ac`, which is 2x the area of the triangle
        ab.cross(ac).length() / 2.0
```

(Someone please double check this)

Reply:
I'm pretty sure `ab` and `ac` should be like this:
```rust
let ab = b - a;
let ac = c - a;
```
Otherwise, that should be correct.

However, I believe a lot of these "textbook formulae" can have some numerical issues especially for thin needle-like triangles. Another alternative is Kahan's formula shown in this paper: https://people.eecs.berkeley.edu/~wkahan/Triangle.pdf

$A = \frac{1}{4} \sqrt{(a+(b+c))(c-(a-b))(c+(a-b))(a+(b-c))}$, where $a \geq b \geq c$.

According to the paper, it seems to have much better accuracy than e.g. Heron's formula and presumably the vector cross product approach. In code, it looks like this:

```rust
let ab = a.distance(b);
let bc = b.distance(c);
let ac= a.distance(c);

// Get side lengths a, b, and c so that a >= b >= c (double-check this)
let (a, b, c) = if ab >= bc {
    if ab >= ac {
        if bc >= ac {
            (ab, bc, ac)
        } else {
            (ab, ac, bc)
        }
    } else {
        (ac, ab, bc)
    }
} else if ab < ac {
    if bc >= ac {
        (bc, ac, ab)
    } else {
        (ac, bc, ab)
    }
} else {
    (bc, ab, ac)
}

let area = 0.25 * ((a + (b + c)) * (c - (a - b)) * (c + (a - b)) * (a + (b - c))).max(0.0).sqrt();
```

This is actually what [Parry uses](https://github.com/dimforge/parry/blob/e57762fe55d4dc5e7799d385e712e779d5884663/src/shape/triangle.rs#L331). Of course, it does have some square roots like Heron's formula, but should be more accurate based on the paper. I haven't verified this myself though.

Reply:
I just skimmed it quickly, so I'm not sure, but the paper seems to be focusing on the task of computing the area knowing only the side lengths, which rules out the cross product from comparison no?

I can see the reasoning that needle-like triangles could be a numerical issue (due to the subtractions inside the cross product), but I'd like to see a demonstration.

Reply:
Hmm, could be. I would expect Parry to have a good reason for using that approach instead of the cross product approach though, considering the extra square roots are a bit expensive. I'm fine with any approach for now though, and we can investigate further in a follow-up.

Reply:
You're right, sorry, that's the outcome of coding without tests or decent researches. Fixed to use `ab.cross(ac).length() / 2.0`. I'll not resolve this conversation because I need to check the paper.

Reply:
I read it quickly and checked some triangles that look like needles. All of them passed. I think my tests might not be good enough. Also, I'll use Heron's Formula, and we can discuss other ways to improve this calculation on another issue.